src/app/clusters/energy-preference-server/energy-preference-server.cpp

```diff
@@ -0,0 +1,221 @@
+/**
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "energy-preference-server.h"
+
+#include <app/util/af.h>
+
+#include <app/util/attribute-storage.h>
+
+#include <app-common/zap-generated/attributes/Accessors.h>
+#include <app-common/zap-generated/callback.h>
+#include <app-common/zap-generated/cluster-objects.h>
+#include <app-common/zap-generated/ids/Attributes.h>
+#include <app/CommandHandler.h>
+#include <app/ConcreteAttributePath.h>
+#include <app/ConcreteCommandPath.h>
+#include <app/util/error-mapping.h>
+#include <lib/core/CHIPEncoding.h>
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::EnergyPreference;
+using namespace chip::app::Clusters::EnergyPreference::Structs;
+using namespace chip::app::Clusters::EnergyPreference::Attributes;
+
+using imcode = Protocols::InteractionModel::Status;
+
+namespace {
+
+class EnergyPrefAttrAccess : public AttributeAccessInterface
+{
+public:
+    EnergyPrefAttrAccess() : AttributeAccessInterface(Optional<EndpointId>::Missing(), EnergyPreference::Id) {}
+
+    CHIP_ERROR Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder) override;
+    CHIP_ERROR Write(const ConcreteDataAttributePath & aPath, AttributeValueDecoder & aDecoder) override;
+};
+
+EnergyPrefAttrAccess gEnergyPrefAttrAccess;
+EnergyPreferenceDelegate * gsDelegate = nullptr;
+
+CHIP_ERROR EnergyPrefAttrAccess::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+{
+    VerifyOrDie(aPath.mClusterId == EnergyPreference::Id);
+    EndpointId endpoint = aPath.mEndpointId;
+    uint32_t ourFeatureMap;
+    bool balanceSupported = (FeatureMap::Get(aPath.mEndpointId, &ourFeatureMap) == EMBER_ZCL_STATUS_SUCCESS) &&
+        ((ourFeatureMap & to_underlying(Feature::kEnergyBalance)) != 0);
+    bool lowPowerSupported = (ourFeatureMap & to_underlying(Feature::kLowPowerModeSensitivity)) != 0;
+
+    switch (aPath.mAttributeId)
+    {
+    case EnergyBalances::Id: {
+        if (balanceSupported == false)
```

Comment:
use `!` instead of ==false.

Reply:
Or, you know, booleans are boolean.. You'd never write `if (foo == true)` but rather if `if (foo)`